relay-server/src/services/project_cache.rs

```diff
@@ -913,6 +895,48 @@ impl ProjectCacheBroker {
         }
     }
 
+    /// Returns backoff timeout for an unspool attempt.
+    fn next_unspool_attempt(&mut self) -> Duration {
+        self.config.spool_envelopes_unspool_interval() + self.buffer_unspool_backoff.next_backoff()
+    }
+
+    fn schedule_unspool(&mut self) {
+        if self.buffer_unspool_handle.is_idle() {
+            // Set the time for the next attempt.
+            let wait = self.next_unspool_attempt();
+            self.buffer_unspool_handle.set(wait);
+        }
+    }
+
+    /// Iterates the buffer index and tries to unspool the envelopes if the project state is still
+    /// valid.
+    ///
+    /// This makes sure we always moving the unspool forward, even if we do not fetch the project
+    /// states updates, but still can process data based on the existing cache.
+    fn handle_periodic_unspool(&mut self) {
+        self.buffer_unspool_handle.reset();
+        let keys = self.index.keys().cloned().collect::<Box<[_]>>();
```

Comment:
How big do we expect the index to be?

Reply:
It's actually tricky, it doesn't work in the loop below because we need to hold a reference to the keys and then call `dequeue` which needs an exclusive reference to `self` (it also modifies the map we'd be iterating over). So without rewriting the `dequeue` method we have to clone all keys here.

Reply:
yeah. I have to get the copy of the project keys here, so I could iterate over them to schedule the dequeue.